### ISSUE TITLE
Relatively poor GPU utilization on Titan X's w/ one-shot examples

### ISSUE THREAD

**Issue by proteneer** [deepchem/deepchem#316]
Running nvidia-smi while training shows a utilization of around 20%, which seems quite poor compared to what we had once achieved with OpenMM (~80%). Note that this is with tox_siamese_one_fold with n_pos and n_neg = 10. Again, nvidia-smi is just a heuristic, but I'm quite curious to know what are the reasons for the low utilization.

Given that the memory usage is about 11.8GB out of 12GB, maybe we're memory bound?

Edit: This is with Titan X Pascal*

**Comment by rbharath** [comment]
@proteneer That's really interesting. The memory usage is much higher than I would have expected, but I'm not that surprised about the low GPU utilization. The one-shot models do a lot of CPU processing (they sample a task to train on, sample a random support set and query set from the data associated from that task, and then do a gradient descent step). My current guess would be that caching some of the CPU computation could possibly improve GPU utilization.

**Comment by proteneer** [comment]
GPU vs CPU benchmarks

(without cuDNN)

fit took 1010.1133031845093 seconds
feed_total: 257.94830989837646
run_total: 747.1647624969482

(with cuDNN, which seems to do nothing)

fit took 997.0082597732544 seconds
feed_total: 255.94153475761414
run_total: 736.102445602417

(CPU-only)

fit took 2788.1415178775787 seconds
feed_total: 210.46477627754211
run_total: 2572.7004868984222
EVAL

so all in all we're looking at a 2.7x speed-up

**Comment by rbharath** [comment]
@proteneer Ah, these numbers make the story clearer. The CPU time isn't really the core blocker `feed_total` contains most of the CPU code. I think what's going on is that our tensorflow code isn't optimized to make best use of the GPU. The graph convolutions and residual LSTMs are complex models that can't make use of cuDNN or other optimized primitives. I think utilization could likely be improved by profiling the GPU usage of these primitives.

**Comment by proteneer** [comment]
Thanks - that makes a lot of sense.

**Comment by peastman** [comment]
I've been profiling the `tox_res_one_fold.py` example, and the main problem is that it's launching huge numbers of tiny kernels that each run for only a few microseconds.  It spends less time executing the kernels than in overhead launching them.

A secondary problem is that it spends about 25% of its time building feed dicts.  If we can find a faster way to do that, it could help some.  Using queues (see #429) might also let us overlap that with GPU computation.

**Comment by rbharath** [comment]
The one-shot models aren't supported in the current version of DeepChem so going to close out this old issue.
